### PR TITLE
chore: new scaffold command for creating resources/data sources

### PR DESCRIPTION
**Pull request by AgustinBettati**
## Description

Link to any related issue(s): [CLOUDP-214957](https://jira.mongodb.org/browse/CLOUDP-214957)

**Summary**: A new make command was defined to scaffold go files associated to a new resource, singular data source, or plural data source. This will speed up development and ensure consistency in the files and code structure of new implementations.

As part of reviewing I suggest running the command locally to see the files that are generated and the code guidelines that are defined.

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [x] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contribution guidelines](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/CONTRIBUTING.md)
- [x] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code
- [x] If changes include deprecations or removals, I defined an isolated PR with a relevant title as it will be used in the auto-generated changelog.

### CODE REVIEW

**Comment by marcosuma** [comment]
This looks **really cool!** I am sure we can further improve it, but LGTM what you have here. Nice job!

**Review by marcosuma — APPROVED** [review]

**Review by oarbusi — APPROVED** [review]

File: tools/scaffold/template/pluraldatasource.tmpl [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+	"github.com/hashicorp/terraform-plugin-framework/datasource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/mongodb/terraform-provider-mongodbatlas/internal/config"
+	"github.com/mongodb/terraform-provider-mongodbatlas/internal/common/dsschema"
+)
+
+var _ datasource.DataSource = &{{.NameCamelCase}}sDS{}
```
Comment (lantoli) [inline tools/scaffold/template/pluraldatasource.tmpl:13]: in the general case s will work for plural resources, for example it doesn't work for data_source_search_indexes.
don't know if allow an optional param that defaults to the name with s,  or maybe we don't need this at the moment and we can change it later.

File: tools/scaffold/scaffold.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+	PluralDataSourceCmd = "plural-data-source"
+)
+
+// struct which is applied to go template files
+type ScaffoldParams struct {
+	GenerationType    string
+	NamePascalCase    string
+	NameCamelCase     string
+	NameSnakeCase     string
+	NameLowerNoSpaces string
```
Comment (andreaangiolillo) [inline tools/scaffold/scaffold.go:25]: Are these properties exported? If not, can we make them private? (lowercase)

**Review by andreaangiolillo — APPROVED** [review]
LGTM

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -47,6 +62,26 @@ For more explained information about plugin override check [Development Override
 - Commit and push your changes to your branch then submit a pull request against the `master` branch.
 - A repo maintainer will review the your pull request, and may either request additional changes or merge the pull request.
 
+#### PR Title Format
+We use [*Conventional Commits*](https://www.conventionalcommits.org/):
```
Comment (lantoli) [inline CONTRIBUTING.md:66]: knitpit: we really use conventional commits only for the commits in master that are normally derived from the PR title, commits in a PR can have any format and they'll be squashed when merged to master.

File: tools/scaffold/scaffold.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+			},
+			{
+				TemplatePath: "tools/scaffold/template/model_test.tmpl",
+				OutputPath:   fmt.Sprintf("%s/model_%s_test.go", folderPath, params.NameSnakeCase),
+			},
+		}, nil
+	case DataSourceCmd:
+		return []FileGeneration{
+			{
+				TemplatePath: "tools/scaffold/template/datasource.tmpl",
```
Comment (lantoli) [inline tools/scaffold/scaffold.go:82]: it'd be nice to generate also a test file for resources / data sources, even mostly empty.

**Review by lantoli — APPROVED** [review]
great job!

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -47,6 +62,26 @@ For more explained information about plugin override check [Development Override
...
+- `fix: description of the PR`: a commit of the type fix patches a bug in your codebase (this correlates with PATCH in Semantic Versioning).
+- `chore: description of the PR`: the commit includes a technical or preventative maintenance task that is necessary for managing the product or the repository, but it is not tied to any specific feature or user story (this correlates with PATCH in Semantic Versioning).
```
Comment (andreaangiolillo) [inline CONTRIBUTING.md:68]: knitpit: I would update the title of this PR to be `chore:` so that the label is removed that is not added to the enhancement section of the changelog. This is a feature for us but not for customers

File: tools/scaffold/scaffold.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+func filesToGenerate(params *ScaffoldParams) ([]FileGeneration, error) {
+	folderPath := fmt.Sprintf("internal/service/%s", params.NameLowerNoSpaces)
+
+	switch params.GenerationType {
+	case ResourceCmd:
+		return []FileGeneration{
+			{
+				TemplatePath: "tools/scaffold/template/resource.tmpl",
```
Comment (lantoli) [inline tools/scaffold/scaffold.go:63]: I think at the moment it is great to run the tool with go run and having the templates in a folder.
in the future if we want to create a binary for the tool we may want to use "embed" to have all template files in the binary

File: tools/scaffold/scaffold.go (hunk 2 above)
Comment (AgustinBettati) [reply tools/scaffold/scaffold.go:25]: verified that properties actually have to be exported when used in go templates.

File: CONTRIBUTING.md (hunk 5 above)
Comment (AgustinBettati) [reply CONTRIBUTING.md:68]: yeah good point, adjusting.

File: CONTRIBUTING.md (hunk 3 above)
Comment (AgustinBettati) [reply CONTRIBUTING.md:66]: Just to clarify this change was just moving the section from one place to another. I agree with what you mention, however the title does make reference to `PR Title Format` which I believe is correct.

File: tools/scaffold/scaffold.go (hunk 4 above)
Comment (AgustinBettati) [reply tools/scaffold/scaffold.go:82]: sounds good, added template file for acceptance tests

File: tools/scaffold/template/pluraldatasource.tmpl (hunk 1 above)
Comment (AgustinBettati) [reply tools/scaffold/template/pluraldatasource.tmpl:13]: yeah I believe given it is an edge case, would not include additional logic at the moment

File: CONTRIBUTING.md (hunk 3 above)
Comment (lantoli) [reply CONTRIBUTING.md:66]: fair point
